File: src/SampleCRM.Web/Services/ProductsService.cs
using OpenRiaServices.DomainServices.Hosting;
using OpenRiaServices.DomainServices.Server;
using SampleCRM.Web.Attributes;
using SampleCRM.Web.Models;
using System;
using System.Data.Entity.Migrations;
using System.Linq;

namespace SampleCRM.Web
{
    [EnableClientAccess]
    public class ProductsService : SampleCRMService
    {
        [Query]
        public IQueryable<Products> GetProducts(string search) =>
            _context.Products
                    .Where(x => x.Name.ToLower().Contains(search.ToLower())
                            || search == "")
                    .OrderBy(c => c.Name);

        public IQueryable<Products> GetProductsCombo() =>
            GetProducts(string.Empty)
                .ToList()
                .Select(x => new Products
                {
                    ProductID = x.ProductID,
                    Name = x.Name
                })
                .AsQueryable();

        [Query]
        public IQueryable<Products> GetProductsWithoutPictures(string search) =>
            GetProducts(search)
                .ToList()
                .Select(x => new Products
                {
                    ProductID = x.ProductID,
                    Name = x.Name,
                    CategoryID = x.CategoryID,
                    Color = x.Color,
                    CreatedOnUTC = x.CreatedOnUTC,
                    DealerPrice = x.DealerPrice,
                    Description = x.Description,
                    Discount = x.Discount,
                    DiscountEndDateUTC = x.DiscountEndDateUTC,
                    DiscountStartDateUTC = x.DiscountStartDateUTC,
                    LastModifiedOnUTC = x.LastModifiedOnUTC,
                    ListPrice = x.ListPrice,
                    SafetyStockLevel = x.SafetyStockLevel,
                    SearchTerms = x.SearchTerms,
                    Size = x.Size,
                    StockUnits = x.StockUnits,
                    TaxType = x.TaxType,
                    Picture = new byte[] { 0 }
                })
                .AsQueryable();

        [Query]
        public IQueryable<Products> GetTopSaleProducts(int limit) =>
            GetProducts(string.Empty)
                .Join(_context.OrderItems, p => p.ProductID, oi => oi.ProductID, (p, oi) => new { Product = p, OrderItem = oi })
                .Join(_context.Orders, oi => oi.OrderItem.OrderID, o => o.OrderID, (oi, o) => new { oi.Product, o.OrderID })
                .GroupBy(po => po.Product)
                .OrderByDescending(group => group.Count())
                .Take(5)
                .Select(group => group.Key)
                .ToList()
                .Select(x => new Products
                {
                    ProductID = x.ProductID,
                    Name = x.Name,
                    ListPrice = x.ListPrice,
                    Picture = new byte[] { 0 }
                })
                .AsQueryable();

        public Products GetProductById(string productId) =>
            GetProducts(string.Empty).SingleOrDefault(x => x.ProductID == productId);

        public byte[] GetProductPicture(string productId)
        {
            var product = GetProductById(productId);
            return product != null ? product.Picture : throw new ArgumentNullException($"No Such Product {productId}");
        }

        [Delete]
        [RestrictAccessReadonlyMode]
        public void DeleteProduct(Products product)
        {
            var dProduct = _context.Products.FirstOrDefault(x => x.ProductID == product.ProductID);
            if (dProduct == null)
                return;

            _context.Products.Remove(dProduct);
            _context.SaveChanges();
        }

        [Insert]
        [RestrictAccessReadonlyMode]
        public void InsertProduct(Products product)
        {
            product.ProductID = new Random().Next((int)Math.Pow(10, 12), (int)Math.Pow(10, 13) - 1).ToString();
            product.CreatedOnUTC = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            product.LastModifiedOnUTC = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            _context.Products.Add(product);
#if DEBUG
            var validationResult = _context.Entry(product).GetValidationResult();
            if (validationResult.ValidationErrors.Any())
            {
                Console.WriteLine($"Validation Error in InsertProduct: {validationResult.ValidationErrors.FirstOrDefault().PropertyName} {validationResult.ValidationErrors.FirstOrDefault().ErrorMessage}");
            }
#endif
            _context.SaveChanges();
        }

        [Update]
        [RestrictAccessReadonlyMode]
        public void UpdateProduct(Products product)
        {
            product.LastModifiedOnUTC = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            _context.Products.AddOrUpdate(product);
            _context.SaveChanges();
        }
    }
}

File: src/SampleCRM.Web/Services/ProductsService.vb
Imports OpenRiaServices.DomainServices.Hosting
Imports OpenRiaServices.DomainServices.Server
Imports SampleCRM.Web.Attributes
Imports System
Imports System.Data.Entity
Imports System.Data.Entity.Migrations
Imports System.Linq

Namespace SampleCRM.Web
    <EnableClientAccess>
    Public Class ProductsService
        Inherits SampleCRMService

        <Query>
        Public Function GetProducts() As IQueryable(Of Products)
            Return _context.Products
        End Function

        <Query>
        Public Function GetTopSaleProducts(ByVal limit As Integer) As IQueryable(Of Products)
            Dim topProducts = _context.Products _
                .Join(_context.OrderItems, Function(p) p.ProductID, Function(oi) oi.ProductID, Function(p, oi) New With {.Product = p, .OrderItem = oi}) _
                .Join(_context.Orders, Function(oi) oi.OrderItem.OrderID, Function(o) o.OrderID, Function(oi, o) New With {.Product = oi.Product, .OrderID = o.OrderID}) _
                .GroupBy(Function(po) po.Product) _
                .OrderByDescending(Function(group) group.Count()) _
                .Take(5) _
                .Select(Function(group) group.Key)
            Return topProducts
        End Function

        Public Function GetProductById(ByVal productId As String) As Products
            Return _context.Products.SingleOrDefault(Function(x) x.ProductID = productId)
        End Function

        <Delete>
        <RestrictAccessDeveloperMode>
        Public Sub DeleteProduct(ByVal product As Products)
            _context.Products.Remove(product)
            Dim dProduct = _context.Products.FirstOrDefault(Function(x) x.ProductID = product.ProductID)
            If dProduct Is Nothing Then Return
            _context.Products.Remove(dProduct)
            _context.SaveChanges()
        End Sub

        <Insert>
        <RestrictAccessDeveloperMode>
        Public Sub InsertProduct(ByVal product As Products)
            product.ProductID = New Random().[Next](CInt(Math.Pow(10, 12)), CInt(Math.Pow(10, 13)) - 1).ToString()
            product.CreatedOn = Date.Now.ToString()
            product.LastModifiedOn = Date.Now.ToString()
            _context.Products.Add(product)
            _context.SaveChanges()
        End Sub

        <Update>
        <RestrictAccessDeveloperMode>
        Public Sub UpdateProduct(ByVal product As Products)
            product.LastModifiedOn = Date.Now.ToString()
            _context.Products.AddOrUpdate(product)
            _context.SaveChanges()
        End Sub
    End Class
End Namespace
